lib/git-fastclone.rb

```diff
@@ -343,17 +345,18 @@ def prefetch(submodule_file)
     # that this repo has been updated on this run of fastclone
     def store_updated_repo(url, mirror, repo_name, fail_hard)
       unless Dir.exist?(mirror)
-        Terrapin::CommandLine.new('git clone', '--mirror :url :mirror')
+        Terrapin::CommandLine.new('git clone', '--mirror :url :mirror' + NO_STDOUT_ERR)
                              .run(url: url.to_s, mirror: mirror.to_s)
       end
 
-      Terrapin::CommandLine.new('cd', ':path; git remote update --prune').run(path: mirror)
+      Terrapin::CommandLine.new('cd', ':path; git remote update --prune' + NO_STDOUT_ERR).run(path: mirror)
 
       reference_updated[repo_name] = true
     rescue Terrapin::ExitStatusError => e
       # To avoid corruption of the cache, if we failed to update or check out we remove
       # the cache directory entirely. This may cause the current clone to fail, but if the
       # underlying error from git is transient it will not affect future clones.
+      puts '[WARN] Removing the fastclone cache.'
       FileUtils.remove_entry_secure(mirror, force: true)
```

Comment:
I wonder if we should be more thoughtful here and only clear the cache directory in the case of certain failures that we know aren't recoverable, given that removing the cache can lead to thundering herd issues for us.

For example, for the recent issue where we had two case insensitive matching branch names, I don't think clearing the cache helped resolve the issue, at least not until the upstream ref has been deleted. We'd have to experiment to see if clearing the cache was actually necessary to resolve the broken state once the bad ref was removed server side.

```
error: cannot lock ref 'refs/heads/<branch>': is at <sha1> but expected <sha2>
```

Reply:
So right now we clear cache only if we consider it as a retriable error, so if we converge the logic here to the one at the end (rescue block for. `with_git_mirror` method) at least for above error we face, we won't clear cache and will just fail hard? I think it's totally reasonable as we have it better than before (clear cache blindly)